### PR TITLE
Add/restore beforeCopy and afterCopy options

### PR DESCRIPTION
**Pull request by Gerych1984**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

1. Add `FileHelper::copyFile` method
2. Add `beforeCopy` and `afterCopy` options to `FileHelper::copyDirectory` and `FileHelper::copyFile`

### CODE REVIEW

File: src/FileHelper.php [hunk 1]
```diff
@@ -412,16 +411,41 @@ public static function copyFile(string $source, string $destination, array $opti
...
+     */
+    private static function processCallback($callback, ...$arguments)
+    {
+        if ($callback === null) {
+            return;
+        }
+
+        if ($callback instanceof Closure) {
+            return $callback(...$arguments);
+        }
```
Comment (vjik) [inline src/FileHelper.php:438]:
```suggestion
        if ($callback instanceof Closure) {
            return $callback(...$arguments);
        }
```

Unnecessary code. Next `call_user_func_array()` can run `Closure`.

File: src/FileHelper.php [hunk 2]
```diff
@@ -412,16 +411,41 @@ public static function copyFile(string $source, string $destination, array $opti
...
+
+        if (is_callable($callback)) {
+            return call_user_func_array($callback, $arguments);
+        }
+
+        $type = \is_object($callback) ? \get_class($callback) : \gettype($callback);
```
Comment (vjik) [inline src/FileHelper.php:444]:
```suggestion
        $type = \is_object($callback) ? \get_class($callback) : \gettype($callback);
```

Use `use` section for functions instead FQN.

File: src/FileHelper.php [hunk 3]
```diff
@@ -335,21 +342,108 @@ public static function copyDirectory(string $source, string $destination, array
...
+     *   beforeCopy?: callable,
+     *   afterCopy?: callable,
+     * } $options
+     *
+     * @return bool
+     */
+    public static function copyFile(string $source, string $destination, array $options = []): bool
+    {
+        if (!is_file($source)) {
+            return false;
```
Comment (vjik) [inline src/FileHelper.php:388]: Throw exception instead of returns `false`. And add test for this case.

File: src/FileHelper.php [hunk 4]
```diff
@@ -335,21 +342,108 @@ public static function copyDirectory(string $source, string $destination, array
...
+     * @psalm-param array{
+     *   dirMode?: int,
+     *   fileMode?: int,
+     *   beforeCopy?: callable,
+     *   afterCopy?: callable,
+     * } $options
+     *
+     * @return bool
+     */
+    public static function copyFile(string $source, string $destination, array $options = []): bool
```
Comment (vjik) [inline src/FileHelper.php:385]:
```suggestion
    public static function copyFile(string $source, string $destination, array $options = []): void
```

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/FileHelper.php (hunk 2 above)
Comment (Gerych1984) [reply src/FileHelper.php:444]: > Use `use` section for functions instead FQN.

Strange. "Fix uses" don't auto-resolve it

File: src/FileHelper.php (hunk 4 above)
Comment (Gerych1984) [reply src/FileHelper.php:385]: But if we need do some action after file copy or not copied. I think it's better like native php`copy` function

Comment (vjik) [reply src/FileHelper.php:385]: Maybe create special exception? And if need action - catch this exception.

File: src/FileHelper.php [hunk 5]
```diff
@@ -335,21 +342,100 @@ public static function copyDirectory(string $source, string $destination, array
...
+        $beforeCopy = $options['beforeCopy'] ?? null;
+
+        if (self::processCallback($beforeCopy, $source, $destination) === false) {
+            return;
+        }
+
+        if (!is_dir($dirname)) {
+            if ($dirMode === null) {
+                throw new RuntimeException('Directory ' . $dirname . ' does not exist and cannot be created.');
+            }
```
Comment (vjik) [inline src/FileHelper.php:402]: Why don't can create directory if not defined dir mode?

Comment (Gerych1984) [reply src/FileHelper.php:402]: > Why don't can create directory if not defined dir mode?

I was try repeat native `copy` function by default

Comment (vjik) [reply src/FileHelper.php:401]: This is change previous behavior: if dir mode not set, then use 0755 (see [here](https://github.com/yiisoft/files/blob/0f0b76ebadd0493d3a30dd257ab6269248439116/src/FileHelper.php#L308)).

File: src/FileHelper.php [hunk 6]
```diff
@@ -335,21 +341,101 @@ public static function copyDirectory(string $source, string $destination, array
...
+            return;
+        }
+
+        if (is_callable($callback)) {
+            return call_user_func_array($callback, $arguments);
+        }
+
+        $type = is_object($callback) ? get_class($callback) : gettype($callback);
+
+        throw new InvalidArgumentException(
```
Comment (vjik) [inline src/FileHelper.php:433]: Add test for this throw exception.

File: tests/FileHelperCallbackTest.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        $this->createFileStructure([
+            $source => $files,
+        ]);
+
+        $basePath = $this->testFilePath;
+        $source = $basePath . '/' . $source;
+        $destination = $basePath . '/test_dst_dir';
+
+        $options = [
+            'beforeCopy' => function (string $source, string $destination) use ($files, $basePath) {
```
Comment (vjik) [inline tests/FileHelperCallbackTest.php:29]: If remove this option - test pass, but must fail.

File: tests/FileHelperCallbackTest.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            $source => $files,
+        ]);
+
+        $basePath = $this->testFilePath;
+        $source = $basePath . '/' . $source;
+        $destination = $basePath . '/test_dst_dir';
+
+        $options = [
+            'beforeCopy' => [$callable, 'beforeCopy'],
+            'afterCopy' => [$callable, 'afterCopy'],
```
Comment (vjik) [inline tests/FileHelperCallbackTest.php:208]: If remove this options - test pass, but must fail.

File: tests/FileHelperCopyFileTest.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+        $this->expectException(InvalidArgumentException::class);
+
+        foreach ($files as $name => $content) {
+            $sourceFile = $source . '/' . $name;
+            $destFile = $destination . '/' . $name;
+
+            FileHelper::copyFile($sourceFile, $destFile);
+
+            $this->assertFileDoesNotExist($destFile);
```
Comment (vjik) [inline tests/FileHelperCopyFileTest.php:85]: Seems, this not need.

**Review by vjik — COMMENTED** [review]
Look's good. Can you also add line to changelog?

**Review by vjik — APPROVED** [review]

**Comment by samdark** [comment]
Thanks.

**Pull request merged**